lens-macro-impl/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+            pub struct #field_name;
+        }
+    });
+
+    // Impl Lens for each field
+    let impls = fields.iter().map(|f| {
+        let field_name = &f.ident;
+        let ty = &f.ty;
+
+        quote! {
+
+            impl Lens<#name, #ty> for #field_name {
+                fn get<'a>(&self, data: &'a #name) -> &'a #ty {
+                    &data.#field_name
+                }
+
+                fn with_mut<V, F: FnOnce(&mut #ty) -> V>(&self, data: &mut #name, f: F) -> V {
+                    f(&mut data.#field_name)
+                }
+            }
+        }
+    });
+
+    let expanded = quote! {
+        pub mod lenses {
+            pub mod #twizzled_name {
+                use super::super::*;
+            
+                use druid::Lens;
+                #(#structs)*
+                #(#impls)*
+            }
+
+        }
+    };
+
+    proc_macro::TokenStream::from(expanded)
+}
+
+struct LensWrapMacroInput {
+    inner_widget: Expr,
+    state: syn::Ident,
+    field: syn::Ident
+}
+
+// TODO: make this less fragile!
+impl Parse for LensWrapMacroInput {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let inner_widget: Expr = input.parse()?;
+        let _comma = <Token![,]>::parse(input)?;
+        let state = syn::Ident::parse(input)?;
+        input.parse::<Token![::]>()?;
+        let field = syn::Ident::parse(input)?;
+
+        // let token_stream = proc_macro2::TokenStream::parse(&input)?;
+        // eprintln!("{:#?}", token_stream);
+
+        Ok(LensWrapMacroInput {
+            inner_widget,
+            state,
+            field
+        })
+    }
+}
+
+// Converts:
+// lens_wrap!(widget, AppState::width)
+// 
+// into
+//
+// druid::LensWrap::new(widget, lenses::AppState::width)
+#[proc_macro_hack]
+pub fn lens_wrap(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let input = parse_macro_input!(input as LensWrapMacroInput);
+
+    let inner_widget = input.inner_widget;
+    let state_name = input.state;
+    let field = input.field;
+
+    let twizzled_name = if is_camel_case(&state_name.to_string()) {
+        let temp_name = to_snake_case(&mut state_name.to_string());
+        Ident::new(&temp_name, Span::call_site())
+    } else {
+        unimplemented!();
+    };
+
+    let expanded = quote! {
+        druid::LensWrap::new(#inner_widget, lenses::#twizzled_name::#field)
+    };
+
+    proc_macro::TokenStream::from(expanded)
+}
+
+//I stole these from rustc!
+fn char_has_case(c: char) -> bool {
+    c.is_lowercase() || c.is_uppercase()
+}
+
+fn is_camel_case(name: &str) -> bool {
+    let name = name.trim_matches('_');
+    if name.is_empty() {
+        return true;
+    }
+
+    // start with a non-lowercase letter rather than non-uppercase
+    // ones (some scripts don't have a concept of upper/lowercase)
+    !name.chars().next().unwrap().is_lowercase()
+        && !name.contains("__")
+        && !name.chars().collect::<Vec<_>>().windows(2).any(|pair| {
```

Comment:
A fancy zero-allocation way to do this would be with `peek`, but this is totally fine as is.

Reply:
another rustc PR!